### ISSUE TITLE
Make public API surface larger

### ISSUE THREAD

**Issue by RemiBou** [dotnet/aspnetcore#19287]
### Is your feature request related to a problem? Please describe.

A lot of classes / methods / fields are made internal and this greatly reduce extension and 3rd party library creation. Only the part used in the documentation/sample/template are made public. There is the same problem with some dependency being created explicitly without DI or without instance public exposition. If someone want to extend this they have to choose one of those bad solution :
- copy/paste a lot of code
- rely on reflection
- try to recreate the same behavior 
- give up. 

I create this issue because I gave up a few times lately:
- Wanted to read the component hierarchy : can't because renderer are not public, we need to rely on people adding correct CascagingParameter
- Wanted to profile component rendering ? same reason as above
- Wanted to profile Http call automatically with MiniProfiler : can't because there is no extension point in HttpClient, HttpMessageinvoker or HttpMessageHandler. 
- Wanted to create a new router that would allow sub-route like router-outlet in angular ? Can"t because there is no public class (most of them are static) in Blazor routing.

I understand making method public is a burden on support and will increase the chance of breaking change, but I think this would be great for the community. And, IMHO, currently the balance is more on the maintainer side than on the community side.

I also understand that this issue does not provide any details and is more a general opinion (maybe my example are bad because I spend only 2 hrs per week looking at this code), but I think this should be something very important to take into account. 

Thanks for your work BTW, this framework is still awesome.

**Comment by javiercn** [comment]
@RemiBou thanks for contacting us.

There are quite a bunch of asks in this issue, I would recommend filing separate issues as that would help us triage and decide independently.

> * Wanted to read the component hierarchy : can't because renderer are not public, we need to rely on people adding correct CascagingParameter.

What specifically are you trying to accomplish here?

> Wanted to profile component rendering ? same reason as above

This should be doable with your own component base class.

> * Wanted to profile Http call automatically with MiniProfiler : can't because there is no extension point in HttpClient, HttpMessageinvoker or HttpMessageHandler.

You can register your own HTTP Client instance in DI and set it up however best suits your needs.

> * Wanted to create a new router that would allow sub-route like router-outlet in angular ? Can"t because there is no public class (most of them are static) in Blazor routing.

I understand the frustration here, but the reality is that making some of these things public will lock us down into a specific API and will hurt future feature work.

I get that this likely what you want to hear, but the reality is that every API that we ship comes has an associated cost in terms of maintenance and our ability to change things later, so we try not to make some APIs public until we are sure the design is factored in a way that allows us to evolve it in the future. Exposing some API we haven't fully thought trough comes at the risk of limiting the evolution of the area in the future.

We take extensibility very seriously, but at the same time we need to balance the time it takes to get it right with the time we have to deliver other features that might add more value.

Hope this helps.

**Comment by RemiBou** [comment]
> This should be doable with your own component base class.
I did it like this : https://github.com/RemiBou/dotnet/blob/master/src/MiniProfiler.Blazor/MiniProfilerComponentBase.cs

```cs
 public class MiniProfilerComponentBase : ComponentBase
    {
        private Timing _currentTiming;

        protected override bool ShouldRender()
        {
            _currentTiming = MiniProfiler.Current.Step("Rendering " + this.GetType().Name);
            return true;
        }
        protected override void OnAfterRender(bool firstRender)
        {
            _currentTiming?.Stop();
        }
    }
```

But I would have prefered to override StateHasChanged or handling this in the Renderer directly so it would give my user a better experience.

> You can register your own HTTP Client instance in DI and set it up however best suits your needs.

I tried but failed : if I inherit from HttpClient, SendAsync is no longer virtual (from HttpMessageInvoker), so I can't decore it. httpMessageHandler is private and readonly, so I can't decore it. I tried to create one with "new" on each method, but that would mean changing all my users HttpClient usage and they won't be able to use HttpClient extension method or any lib that uses HttpClient because it would call HttpClient implementation instead of my own. See my (lame) attempt here : https://github.com/RemiBou/dotnet/blob/2976772c88582becc1b5310a522a7eefcc27e4a8/src/MiniProfiler.Blazor/MiniProfilerBlazorHttpClient.cs

> I understand the frustration here, but the reality is that making some of these things public will lock us down into a specific API and will hurt future feature work.

Thank you for your answer, I admit it's a broad issue and might be better treated as different issues.  I tried to open an issue (https://github.com/dotnet/aspnetcore/issues/18500) we'll see how it goes and I might open more in the future.

My issue is really about this frustration I meet on nearly every attempts to extend ASPNET Core (or even net core), and the "balance" that might nee to be thought about. If I give up on these 3-4 occasion (maybe some more), how many other people gave up ? How many 3rd part library never existed because of this (I know HttpClient profiling is a PITA for MiniProfiler maintainer : https://twitter.com/Nick_Craver/status/1089503751570931713) ? 

At my current company I use Spring Boot, and honestly so far every extension attempt was a success because nearly every method are public or protected (and there is no virtual in Java) and most of the dependencies are created via DI. This is a purely subjective opinion with no data for backing it up but I wish you'd consider the current balance and take it into account when there are discussions about the fact that dotnet is Microsoft centered.

Thanks for reading :)
